### PR TITLE
Hashcode and some equals methods

### PR DESCRIPTION
**Pull request by SimonIT**
I have to use `Arrays.hashCode` instead of `Objects.hash` because it's only available on android api level 19 an higher @jeremyz

### CODE REVIEW

File: core/src/ch/asynk/gdx/boardgame/Path.java [hunk 1]
```diff
@@ -96,4 +97,20 @@ public boolean nextPosition(Piece piece, Vector3 v)
         to = null;
         return true;
     }
+
+    @Override public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        if (!super.equals(o)) return false;
+        Path tiles = (Path) o;
```
Comment (jeremyz) [inline core/src/ch/asynk/gdx/boardgame/Path.java]: shouldn't you compare tiles in between from and to too

File: core/src/ch/asynk/gdx/boardgame/Tile.java [hunk 2]
```diff
@@ -41,4 +41,19 @@ public void enableOverlay(int i, boolean enable)
...
     }
+
+    @Override public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Tile tile = (Tile) o;
+        return Float.compare(tile.x, x) == 0 &&
+                Float.compare(tile.y, y) == 0 &&
+                (overlays == tile.overlays) || (overlays != null && overlays.equals(tile.overlays));
```
Comment (jeremyz) [inline core/src/ch/asynk/gdx/boardgame/Tile.java]: overlays shouldn't be used here, a tile is defined by it's coordinates only.

Comment (SimonIT) [reply core/src/ch/asynk/gdx/boardgame/Tile.java]: Nice idea, but how would you do that?

**Comment by SimonIT** [comment]
So what should I do? I need this functionality @jeremyz

**Comment by jeremyz** [comment]
it would be very nice if you could split these in different PR,
 - one with the equals() and hashcode() (for what do you need these ?!?)
 - one with the cast corrections
 - one with the protected -> public

I don't think you should reorder the import statements,
I've tried to be consistent the whole project, import java.* first then the externallibs.* then at last localproject.*
It's better to keep that logic, unless you have a good reason for that.

**Comment by SimonIT** [comment]
Ok, I do everything 👍🏽
My idea with hashcode() is to detect wether the children of the assembly changed and computeGeometries() then.
IntelliJ changed the imports, I can take the order you want.

**Comment by SimonIT** [comment]
Ok, I do everything 👍🏽
My idea with hashcode() is to detect wether the children of the assembly changed and computeGeometries() then.
IntelliJ changed the imports, I can take the order you want.

**Comment by jeremyz** [comment]
Maybe there is something missing,
but you don't need that, just set tainted = true,
look at Element and Label, each method that changes something set tainted=true,
then in Label.draw() computeGeometry() is called if tainted,
it's all automatic

**Comment by SimonIT** [comment]
But what if the size of the Label changed by reference, so my custom assembly widget has to recompute its geometries but doesn't know that one of its children changed. So I check if the hashcode changed and then recompute its geometries

**Comment by jeremyz** [comment]
Using a trick such as tracking children hash code changes to recompute geometry is way too much.
(it's a duplicate of the tainted mechanism, and needs lot of changes if a new trait is added)
The correct way would be to propagate the tainted flag change to the parent, all the way up the tree,
but from the start I thought that recomputing geometry all over the UI tree for every change is inefficient. It's not a real UI framework, just some basic widgets, that are supposed to be shown in a static way.

I'm trying hard to remember why I added that Assembly class, It's not part of the original design ...
I remember it had something to do with Root, but was never intended to hold elements together as you would do with a Panel or Window (yep, you told me about documentation).

What you should do if you design a custom widget is what I did for the Menu class.
The widget owns the references, and if you want to manipulate what's behind the reference, say a Label, add a method to the widget to do that, and taint the widget in the same time that you modify the Label.

What do you think ?

**Comment by jeremyz** [comment]
We could add a configurable tainted propagation up the tree ?
```
public class Element {
   …
   public boolean taintParent = false;
   …
   public void taint() {
      this.tainted = true;
      if (parent != null && taintParent) parent.taint();
   }
}
```
would it fix you issue ?

**Comment by SimonIT** [comment]
Good idea, I think so
Should I make a PR or do you commit directly? @jeremyz 
btw, I need #6

**Comment by jeremyz** [comment]
I just pushed the taintParent stuff, let me know if it fulfills your needs.

**Comment by SimonIT** [comment]
I will, I only need then the public computeGeometry #6

**Pull request closed without merging**